src/native/libs/System.Native/pal_runtimeinformation.c

```diff
@@ -48,8 +53,93 @@ int32_t SystemNative_GetUnixVersion(char* version, int* capacity)
     return 0;
 }
 
-/* Returns an int representing the OS Architecture. -1 if same as process architecture. */
+// Keep in sync with System.Runtime.InteropServices.Architecture enum
+enum
+{
+    ARCH_X86,
+    ARCH_X64,
+    ARCH_ARM,
+    ARCH_ARM64,
+    ARCH_WASM,
+    ARCH_S390X,
+    ARCH_LOONGARCH64,
+    ARCH_ARMV6,
+};
+
 int32_t SystemNative_GetOSArchitecture()
 {
-    return -1;
+#ifdef TARGET_WASM
+    return ARCH_WASM;
+#else
+    int32_t result = -1;
+#ifdef TARGET_SUNOS
+    // On illumos/Solaris, the recommended way to obtain machine
+    // architecture is using `sysinfo` rather than `utsname.machine`.
+
+    char isa[32];
+    if (sysinfo(SI_ARCHITECTURE_K, isa, sizeof(isa)) > -1)
+    {
+        {
+#else
+    {
+        struct utsname _utsname;
+        if (uname(&_utsname) > -1)
+        {
+            char* isa = _utsname.machine;
+#endif
+            // aarch64 or arm64: arm64
+            if (strcmp("aarch64", isa) == 0 || strcmp("arm64", isa) == 0)
+            {
+                result = ARCH_ARM64;
+            }
+
+            // starts with "armv6" (armv6h or armv6l etc.): armv6
+            else if (strncmp("armv6", isa, strlen("armv6")) == 0)
+            {
+                result = ARCH_ARMV6;
+            }
+
+            // starts with "arm": arm
+            else if (strncmp("arm", isa, strlen("arm")) == 0)
+            {
+                result = ARCH_ARM;
+            }
+
+            // x86_64 or amd64: x64
+            else if (strcmp("x86_64", isa) == 0 ||
+                strcmp("amd64", isa) == 0)
+            {
+#ifdef TARGET_OSX
+                int is_translated_process = 0;
+                size_t size = sizeof(is_translated_process);
+                if (sysctlbyname("sysctl.proc_translated", &is_translated_process, &size, NULL, 0) == 0)
```

Comment:
Do we need to check what the API actually returned in `is_translated_process` ?

Reply:
I tested it in a standalone program: http://sprunge.us/nJ9fvU. Here are the results:

```zsh
# native
% cc -DTARGET_OSX x.c
% file ./a.out       
./a.out: Mach-O 64-bit executable arm64
% ./a.out            
arm64 (3)

# emulated
% cc -DTARGET_OSX -arch x86_64 x.c
% file ./a.out                    
./a.out: Mach-O 64-bit executable x86_64
% ./a.out                         
arm64 (3)
```

I wanted to try it with corerun, but I was having issues with that:

1. `./build.sh -a x64` or `./build.sh -a x64 -cross` on M1 Pro was failing during the cross component compilation. So I built libs, host and clr without cross components. Then generated layout with `src/tests/build.sh -generatelayoutonly -p:LibrariesConfiguration=Debug -x64`.
2. running `CORE_ROOT=$(pwd)/../runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root ../runtime/artifacts/tests/coreclr/OSX.x64.Debug/Tests/Core_Root/corerun bin/Debug/net7.0/osx-x64/testapp1.dll` failed with `coreclr_initialize failed - Error: 0x80004005`.